### ISSUE TITLE
Indices on Fields in Composite Types: Client & Query Engine

### ISSUE THREAD

**Issue by pimeys** [prisma/prisma#12409]
A unique or fulltext index in a model can point to a field that is in a singular composite type:

``` prisma
type FooFoo {
  bar Int
}

model Foo {
  id  Int    @id
  foo FooFoo

  @@unique([foo.bar])
}
```

Or it can point to a field in an array of composite types:

``` prisma
type FooFoo {
  bar Int
}

model Foo {
  id  Int      @id
  foo FooFoo[]

  @@unique([foo.bar])
}
```

The composite type field can be just a part of the index which includes fields directly from the model too.

We should enable `findUnique` operations, and fulltext operations to these fields if corresponding indices are defined correctly.

If the unique index points to an array of composite types, with the following data:

```json
{
  "name": "Musti",
  "addresses": [{ "city": "Berlin" }, { "city": "Helsinki" }]
}
```

And our PSL says:

``` prisma
type Address {
  city String
}

model User {
  name      String
  addresses Address[]

  @@unique([addresses.city])
}
```

All of these writes will lead to unique constraint violation:

```json
{ "name": "Musti", "addresses": [{ "city": "Helsinki" }] }
{ "name": "Musti", "addresses": [{ "city": "Berlin" }] }
{ "name": "Musti", "addresses": [{ "city": "Helsinki" }, { "city": "Berlin" }] }
```

And the client can find the record from the collection with two different queries:

``` javascript
findUniqueRecord({
  where: {
    "name": "Musti",
    "addresses": { "city": "Berlin" }
  }
});

findUniqueRecord({
  where: {
    "name": "Musti",
    "addresses": { "city": "Helsinki" }
  }
});
```

## Dependencies

The [PSL issue](https://github.com/prisma/prisma/issues/12406) must be merged before starting to work with this one.

## Full proposal

[Private document in Notion](https://www.notion.so/prismaio/Proposal-Composite-Type-Indices-03baf2ea1865417684fdab7fda2ee007)

**Comment by matthewmueller** [comment]
This will be released in 4.3.0

**Comment by jkomyno** [comment]
Fixed by https://github.com/prisma/prisma-engines/pull/3107.
